File: Samples/C#/Debugger/MainForm.cs
using System;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

namespace Debugger
{
    public partial class MainForm : Form
    {
        public delegate Object MyDelegate(params Object[] args);

        public MainForm()
        {
            InitializeComponent();
            tbMain.Document.HighlightingStrategy = ICSharpCode.TextEditor.Document.HighlightingManager.Manager.FindHighlighterForFile("file.js");
            tbMain.TextChanged += new EventHandler(tbMain_TextChanged);
            ScriptEngine.RunInThread = true;
            ScriptEngine.Debug = true;
            ScriptEngine.StatusChanged += new EventHandler(ScriptEngine_StatusChanged);
            ScriptEngine.DebugTracePoint += new EventHandler<RemObjects.Script.ScriptDebugEventArgs>(ScriptEngine_DebugTracePoint);
            ScriptEngine.DebugThreadExit += new EventHandler<RemObjects.Script.ScriptDebugEventArgs>(ScriptEngine_DebugThreadExit);
            ScriptEngine.Globals.SetVariable("writeln", new MyDelegate(delegate(Object[] args)
            {
                Invoke(new MyDelegate(delegate(Object[] args2)
                {
                    String s = "";
                    if (args2 != null)
                        s = String.Join(" ", (from x in args2 select x == null ? "" : x.ToString()).ToArray());

                    s += "\r\n";
                    edOutput.AppendText(s);
                    this.tabs.SelectedTab = tbOutput;
                    return null;
                }), new Object[] { args });
                return null;
            }));
        }

        void ScriptEngine_StatusChanged(Object sender, EventArgs e)
        {
            RemObjects.Script.ScriptStatus lStatus = ScriptEngine.Status;
            BeginInvoke(new Action(delegate
            {
                switch (lStatus)
                {
                    case RemObjects.Script.ScriptStatus.Paused:
                        Text = "Script Editor [Paused]";
                        break;

                    case RemObjects.Script.ScriptStatus.Pausing:
                        Text = "Script Editor [Pausing]";
                        break;

                    case RemObjects.Script.ScriptStatus.Running:
                        edOutput.Clear();
                        Text = "Script Editor [Running]";
                        break;

                    case RemObjects.Script.ScriptStatus.StepInto:
                        Text = "Script Editor [Running (Step Into)]";
                        break;

                    case RemObjects.Script.ScriptStatus.StepOut:
                        Text = "Script Editor [Running (Step Out)]";
                        break;

                    case RemObjects.Script.ScriptStatus.StepOver:
                        Text = "Script Editor [Running (Step Over)]";
                        break;

                    case RemObjects.Script.ScriptStatus.Stopping:
                        Text = "Script Editor [Stopping]";
                        break;

                    default:
                        if (ScriptEngine.RunException != null)
                        {
                            edOutput.AppendText("\r\nException: \r\n" + ScriptEngine.RunException.ToString());
                            this.tabs.SelectedTab = tbOutput;
                        }
                        else
                        {
                            edOutput.AppendText("\r\nResult: \r\n" + ScriptEngine.RunResult);
                        }
                        tbMain.Document.MarkerStrategy.RemoveAll(a => (a.TextMarkerType == ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock && a.Color == Color.Red));
                        tbMain.Refresh();
                        Text = "Script Editor"; // done running
                        break;
                }
            }));
        }

        void ScriptEngine_DebugThreadExit(Object sender, RemObjects.Script.ScriptDebugEventArgs e)
        {
            Invoke(new Action(delegate
                {
                    tbMain.Document.MarkerStrategy.RemoveAll(a => (a.TextMarkerType == ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock && a.Color == Color.Red));
                    tbMain.Refresh();
                }));
        }

        void ScriptEngine_DebugTracePoint(Object sender, RemObjects.Script.ScriptDebugEventArgs e)
        {
            if (e.SourceSpan.IsValid)
            {
                Invoke(new Action(delegate
                {
                    tbMain.Document.MarkerStrategy.RemoveAll(a => (a.TextMarkerType == ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock && a.Color == Color.Red));
                    Int32 lStart = tbMain.Document.PositionToOffset(new ICSharpCode.TextEditor.TextLocation(e.SourceSpan.Start.Column - 1, e.SourceSpan.Start.Line - 1));
                    Int32 lEnd = tbMain.Document.PositionToOffset(new ICSharpCode.TextEditor.TextLocation(e.SourceSpan.End.Column - 1, e.SourceSpan.End.Line - 1));
                    tbMain.Document.MarkerStrategy.AddMarker(new ICSharpCode.TextEditor.Document.TextMarker(lStart, lEnd - lStart + 1, ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock, Color.Red));
                    tbMain.Refresh();
                }));
            }
            else
            {
                Invoke(new Action(delegate
                {
                    tbMain.Document.MarkerStrategy.RemoveAll(a => (a.TextMarkerType == ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock && a.Color == Color.Red));
                    tbMain.Refresh();
                }));
            }
            Invoke(new Action(delegate
            {
                lvLocals.Items.Clear();
                for (Int32 i = ScriptEngine.CallStack.Count - 1; i >= 0; i--)
                {
                    var el = ScriptEngine.CallStack[i];
                    ListViewItem item = new ListViewItem("[METHOD]");
                    item.SubItems.Add(el.Method ?? null);
                    lvLocals.Items.Add(item);
                    item = new ListViewItem("this");
                    item.SubItems.Add(el.This == null ? "this" : el.This.ToString());
                    lvLocals.Items.Add(item);

                    foreach (var name in el.Frame.Names())
                    {
                        item = new ListViewItem(name);
                        var val = el.Frame.GetBindingValue(name, false);
                        item.SubItems.Add(val == null ? "(null)" : val.ToString());
                        lvLocals.Items.Add(item);
                    }
                }
            }));
        }

        void tbMain_TextChanged(Object sender, EventArgs e)
        {
            Changed();
        }

        private void stepIntoToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdStepInto();
        }

        private void stopToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdStop();
        }

        private void setClearBreakpointToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdSetBreakpoint();
        }

        private void stepOutToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdStepOut();
        }

        private void stepOverToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdStepOver();
        }

        private void runToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdRun();
        }

        private void debugToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            runToolStripMenuItem.Enabled = IsPaused() || !IsDebugging();
            stepIntoToolStripMenuItem.Enabled = runToolStripMenuItem.Enabled;
            stepOutToolStripMenuItem.Enabled = IsPaused();
            stepOverToolStripMenuItem.Enabled = runToolStripMenuItem.Enabled;
            setClearBreakpointToolStripMenuItem.Enabled = true;
            stopToolStripMenuItem.Enabled = IsDebugging();
        }

        private void newToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdNew();
        }

        private void openToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdOpen();
        }

        private void saveToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdSave();
        }

        private void saveAsToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdSaveAs();
        }

        private void exitToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            cmdQuit();
        }

        private void fileToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            newToolStripMenuItem.Enabled = CanFileCommands();
            openToolStripMenuItem.Enabled = CanFileCommands();
            saveAsToolStripMenuItem.Enabled = CanFileCommands();
            saveToolStripMenuItem.Enabled = CanFileCommands();
            exitToolStripMenuItem.Enabled = CanFileCommands();
        }

        private void MainForm_FormClosing(Object sender, FormClosingEventArgs e)
        {
            if (!CanQuit())
                e.Cancel = true;
        }

        private void aboutToolStripMenuItem_Click(Object sender, EventArgs e)
        {
            CmdAbout();
        }

    }
}

File: Samples/Oxygene/Debugger/MainForm.pas
namespace Debugger;

interface

uses
  System.Drawing,
  System.Collections,
  System.Collections.Generic,
  System.Linq,
  System.Windows.Forms,
  System.ComponentModel;

type
  /// <summary>
  /// Summary description for MainForm.
  /// </summary>
  MainForm = partial class(System.Windows.Forms.Form)
  private
    method MainForm_FormClosing(sender: Object; e: FormClosingEventArgs);
    method ScriptEngine_DebugThreadExit(sender: Object; e: RemObjects.Script.ScriptDebugEventArgs);
    method ScriptEngine_DebugTracePoint(sender: Object; e: RemObjects.Script.ScriptDebugEventArgs);
    method ScriptEngine_StatusChanged(sender: Object; e: EventArgs);
    method aboutToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method debugToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method exitToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method fileToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method newToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method openToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method runToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method saveAsToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method saveToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method setClearBreakpointToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method stepIntoToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method stepOutToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method stepOverToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method stopToolStripMenuItem_Click(sender: Object; e: EventArgs);
    method tbMain_TextChanged(sender: Object; e: EventArgs);
  protected
    method Dispose(aDisposing: Boolean); override;
  public
    constructor;
  end;

  MyDelegate = public block(params args: array of Object);

implementation

{$REGION Construction and Disposition}
constructor MainForm;
begin
  //
  // Required for Windows Form Designer support
  //
  InitializeComponent();

  tbMain.Document.HighlightingStrategy := ICSharpCode.TextEditor.Document.HighlightingManager.Manager.FindHighlighterForFile('file.js');
  tbMain.TextChanged += new EventHandler(tbMain_TextChanged);
  ScriptEngine.RunInThread := true;
  ScriptEngine.Debug := true;
  ScriptEngine.StatusChanged += new EventHandler(ScriptEngine_StatusChanged);
  ScriptEngine.DebugTracePoint += new EventHandler<RemObjects.Script.ScriptDebugEventArgs>(ScriptEngine_DebugTracePoint);
  ScriptEngine.DebugThreadExit += new EventHandler<RemObjects.Script.ScriptDebugEventArgs>(ScriptEngine_DebugThreadExit);
  ScriptEngine.Globals.SetVariable('writeln', new MyDelegate(method (args: array of Object) begin
    Invoke(new MyDelegate(method (args2: array of Object) begin
      var s: String := '';
      if args2 <> nil then      
        s := String.Join(' ', ( from x in args2 select iif(x = nil, '', x.ToString())).ToArray());
      s := s + #13#10;
      edOutput.AppendText(s);
      self.tabs.SelectedTab := tbOutput;
      exit ;
    end), array of Object([args]));
    exit nil;
  end));
end;

method MainForm.Dispose(aDisposing: Boolean);
begin
  if aDisposing then begin
    if assigned(components) then
      components.Dispose();

    //
    // TODO: Add custom disposition code here
    //
  end;
  inherited Dispose(aDisposing);
end;
{$ENDREGION}

method MainForm.ScriptEngine_StatusChanged(sender: Object; e: EventArgs);
begin
  var lStatus: RemObjects.Script.ScriptStatus := ScriptEngine.Status;
  BeginInvoke(new Action(method begin
    case lStatus of 
      RemObjects.Script.ScriptStatus.Paused: begin
        Text := 'Script Editor [Paused]';
      end;
      
      RemObjects.Script.ScriptStatus.Pausing: begin
        Text := 'Script Editor [Pausing]';
      end;
      
      RemObjects.Script.ScriptStatus.Running: begin
        edOutput.Clear();
        Text := 'Script Editor [Running]';
      end;
      
      RemObjects.Script.ScriptStatus.StepInto: begin
        Text := 'Script Editor [Running (Step Into)]';
      end;
      
      RemObjects.Script.ScriptStatus.StepOut: begin
        Text := 'Script Editor [Running (Step Out)]';
      end;
      
      RemObjects.Script.ScriptStatus.StepOver: begin
        Text := 'Script Editor [Running (Step Over)]';
      end;
      
      RemObjects.Script.ScriptStatus.Stopping: begin
        Text := 'Script Editor [Stopping]';
      end
      else begin
        if ScriptEngine.RunException <> nil then begin
          edOutput.AppendText(#13#10'Exception: '#13#10 + ScriptEngine.RunException.ToString);
          self.tabs.SelectedTab := tbOutput;
        end
        else begin
          edOutput.AppendText(#13#10'Result: '#13#10 + ScriptEngine.RunResult);
        end;
        tbMain.Document.MarkerStrategy.RemoveAll(a -> ((a.TextMarkerType = ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock) and (a.Color = Color.Red)));
        tbMain.Refresh;
        Text := 'Script Editor';
      end;
    end;    // done running

  end));
end;

method MainForm.ScriptEngine_DebugThreadExit(sender: Object; e: RemObjects.Script.ScriptDebugEventArgs);
begin
  Invoke(new Action(method begin
    tbMain.Document.MarkerStrategy.RemoveAll(a -> ((a.TextMarkerType = ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock) and (a.Color = Color.Red)));
    tbMain.Refresh;
  end));
end;

method MainForm.ScriptEngine_DebugTracePoint(sender: Object; e: RemObjects.Script.ScriptDebugEventArgs);
begin
  if e.SourceSpan.IsValid then begin
    Invoke(new Action(method begin
      tbMain.Document.MarkerStrategy.RemoveAll(a -> ((a.TextMarkerType = ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock) and (a.Color = Color.Red)));
      var lStart: Int32 := tbMain.Document.PositionToOffset(new ICSharpCode.TextEditor.TextLocation(e.SourceSpan.Start.Column - 1, e.SourceSpan.Start.Line - 1));
      var lEnd: Int32 := tbMain.Document.PositionToOffset(new ICSharpCode.TextEditor.TextLocation(e.SourceSpan.End.Column - 1, e.SourceSpan.End.Line - 1));
      tbMain.Document.MarkerStrategy.AddMarker(new ICSharpCode.TextEditor.Document.TextMarker(lStart, lEnd - lStart + 1, ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock, Color.Red));
      tbMain.Refresh;
    end));
  end
  else begin
    Invoke(new Action(method begin
      tbMain.Document.MarkerStrategy.RemoveAll(a -> ((a.TextMarkerType = ICSharpCode.TextEditor.Document.TextMarkerType.SolidBlock) and (a.Color = Color.Red)));
      tbMain.Refresh;
    end))
  end;
  Invoke(new Action(method begin
    lvLocals.Items.Clear;
    begin
      var i: Int32 := ScriptEngine.CallStack.Count - 1;
      while i >= 0 do begin begin
          var el := ScriptEngine.CallStack[i];
          var item: ListViewItem := new ListViewItem('[METHOD]');
          //item.SubItems.Add(coalesce(el.Method, nil)); <-- Type mismatch
          item.SubItems.Add(el.Method);
          lvLocals.Items.Add(item);
          item := new ListViewItem('this');
          item.SubItems.Add(iif(el.This = nil, 'this', el.This.ToString()));
          lvLocals.Items.Add(item);

          for each name in el.Frame.Names() do begin
            item := new ListViewItem(name);
            var val := el.Frame.GetBindingValue(name, false);
            item.SubItems.Add(iif(val = nil, '(null)', val.ToString()));
            lvLocals.Items.Add(item);
          end
        end;
// TODO: not supported Increment might not get called when using continue
        {POST}dec(i);
      end;
    end;
  end));
end;

method MainForm.tbMain_TextChanged(sender: Object; e: EventArgs);
begin
  Changed;
end;

method MainForm.stepIntoToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdStepInto;
end;

method MainForm.stopToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdStop;
end;

method MainForm.setClearBreakpointToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdSetBreakpoint;
end;

method MainForm.stepOutToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdStepOut;
end;

method MainForm.stepOverToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdStepOver;
end;

method MainForm.runToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdRun;
end;

method MainForm.debugToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  runToolStripMenuItem.Enabled := (IsPaused()) or (not IsDebugging());
  stepIntoToolStripMenuItem.Enabled := runToolStripMenuItem.Enabled;
  stepOutToolStripMenuItem.Enabled := IsPaused;
  stepOverToolStripMenuItem.Enabled := runToolStripMenuItem.Enabled;
  setClearBreakpointToolStripMenuItem.Enabled := true;
  stopToolStripMenuItem.Enabled := IsDebugging;
end;

method MainForm.newToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdNew;
end;

method MainForm.openToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdOpen;
end;

method MainForm.saveToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdSave;
end;

method MainForm.saveAsToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdSaveAs;
end;

method MainForm.exitToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  cmdQuit;
end;

method MainForm.fileToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  newToolStripMenuItem.Enabled := CanFileCommands;
  openToolStripMenuItem.Enabled := CanFileCommands;
  saveAsToolStripMenuItem.Enabled := CanFileCommands;
  saveToolStripMenuItem.Enabled := CanFileCommands;
  exitToolStripMenuItem.Enabled := CanFileCommands;
end;

method MainForm.MainForm_FormClosing(sender: Object; e: FormClosingEventArgs);
begin
  if not CanQuit then    
    e.Cancel := true;
end;

method MainForm.aboutToolStripMenuItem_Click(sender: Object; e: EventArgs);
begin
  CmdAbout;
end;

end.
